### PR TITLE
feat!: add MsgCreatePermanentLockedVestingAccount

### PR DESCRIPTION
**Pull request by clevinson**
ref: https://github.com/regen-network/regen-ledger/issues/188

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/develop/CONTRIBUTING.md#pr-targeting))
- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added entries in `PENDING.md` with issue # 
- [ ] rereviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by clevinson** [comment]
I was reviewing https://github.com/regen-network/regen-ledger/pull/220 to try and make sure I covered everything necessary for amino-json signing. I would appreciate a note if I missed anything. @blushi ?

File: x/auth/vesting/types/codec.go [hunk 1]
```diff
@@ -53,6 +53,7 @@ func RegisterInterfaces(registry types.InterfaceRegistry) {
 	registry.RegisterImplementations(
 		(*sdk.Msg)(nil),
 		&MsgCreateVestingAccount{},
+		&MsgCreatePermanentLockedAccount{},
```
Comment (clevinson) [inline x/auth/vesting/types/codec.go:56]: I didn't end up having to add `MsgCreatePermanentLockedAccount` to the `RegisterLegacyAminoCodec` block up above (line 14). I'm not totally sure why... but I managed to test on my Ledger device with simapp, and I successfully used the CLI for creating & signing new permanent locked accounts.

**Review by ryanchristo — APPROVED** [review]
tACK. Tested with CLI and ledger device and compared with existing create vesting account message and pending create periodic vesting account message. Should this be documented somewhere? Should we have a running changelog for patches we implement on the fork?

**Comment by clevinson** [comment]
I think changeling at minimum is a good idea. Outside of that, personally I think CLI with clear comments / self documenting is probably good enough. I'll add the changelog though.

Afaiu, there isn't any expectation that anybody is making use of this feature aside from Regen Foundation, for the purpose of instantiating CSDAO prototypes.

**Comment by clevinson** [comment]
Turns out the error I was receiving when trying to add the Msg type name for `amino-json` support was that the type name was too long and broke Ledger Nano support...

see https://github.com/regen-network/regen-ledger/issues/468 for more details.  Anyway-- it should all be good now!

File: CHANGELOG.md [hunk 2]
```diff
@@ -35,6 +35,14 @@ Ref: https://keepachangelog.com/en/1.0.0/
 
 # Changelog
 
+<<<<<<< HEAD
```
Comment (amaurym) [inline CHANGELOG.md:38]:
```suggestion
```

File: x/auth/vesting/types/msgs.go [hunk 3]
```diff
@@ -75,3 +80,63 @@ func (msg MsgCreateVestingAccount) GetSigners() []sdk.AccAddress {
...
+	if err != nil {
+		return err
+	}
+	if err := sdk.VerifyAddressFormat(from); err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender address: %s", err)
+	}
+
+	if err := sdk.VerifyAddressFormat(to); err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid recipient address: %s", err)
+	}
```
Comment (technicallyty) [inline x/auth/vesting/types/msgs.go:116]:
```suggestion
	if _, err := sdk.AccAddressFromBech32(msg.FromAddress); err != nil {
	    return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid sender address: %s", err)
	}
	if _, err := sdk.AccAddressFromBech32(msg.ToAddress); err != nil {
	    return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid recipient address: %s", err)
	}
```

no need to call `sdk.VerifyAddressFormat` here - it's already being done in `sdk.AccAddressFromBech32`

File: x/auth/vesting/types/msgs.go [hunk 4]
```diff
@@ -75,3 +80,63 @@ func (msg MsgCreateVestingAccount) GetSigners() []sdk.AccAddress {
...
+func (msg MsgCreatePermanentLockedAccount) GetSignBytes() []byte {
+	return sdk.MustSortJSON(amino.MustMarshalJSON(&msg))
+}
+
+// GetSigners returns the expected signers for a MsgCreatePermanentLockedAccount.
+func (msg MsgCreatePermanentLockedAccount) GetSigners() []sdk.AccAddress {
+	from, err := sdk.AccAddressFromBech32(msg.FromAddress)
+	if err != nil {
+		panic(err)
+	}
```
Comment (technicallyty) [inline x/auth/vesting/types/msgs.go:140]:
```suggestion
	from, _ := sdk.AccAddressFromBech32(msg.FromAddress)
```

the validate basic method will already ensure we have a valid address, no need to check error

**Review by technicallyty — APPROVED** [review]
looks good - pre-approving with some smol nits

**Comment by amaurym** [comment]
Nice work, would really like to see this PR in the SDK at some point too!

File: x/auth/vesting/msg_server.go [hunk 5]
```diff
@@ -98,3 +98,68 @@ func (s msgServer) CreateVestingAccount(goCtx context.Context, msg *types.MsgCre
...
+
+	if acc := ak.GetAccount(ctx, to); acc != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "account %s already exists", msg.ToAddress)
+	}
+
+	baseAccountI := ak.NewAccountWithAddress(ctx, to)
+
+	baseAcc, ok := baseAccountI.(*authtypes.BaseAccount)
+	if !ok {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid account type; expected: BaseAccount, got: %T", baseAccountI)
```
Comment (robert-zaremba) [inline x/auth/vesting/msg_server.go:131]: This can be `sdkerrors.ErrInvalidRequest,Wrapf`.... which simplifies the code

**Pull request merged**